### PR TITLE
Allow `Brewfile` location to be specified in ENV

### PR DESCRIPTION
**Pull request by aramprice**
Adds support for `HOMEBREW_BUNDLE_FILE` as a way to specify the location
of the `Brewfile` to be used by `brew bundle` commands.

Specifying `HOMEBREW_BUNDLE_FILE` when `--global` or `--file` are also
specified results in an error.

Existing behavior wherein `--global` overrides `--file` with no error is
preserved. This could be revisited in the future, the code in
`brewfile.rb` would likely need further refactoring.

* This addresses an issue initially raised in https://github.com/Homebrew/homebrew-bundle/issues/368.

### CODE REVIEW

File: lib/bundle/brewfile.rb [hunk 1]
```diff
@@ -5,19 +5,28 @@ module Brewfile
...
-      if ARGV.include?("--global")
-        Pathname.new("#{ENV["HOME"]}/.Brewfile")
-      else
-        filename = ARGV.value("file")
-        if filename == "-"
-          filename = if dash_writes_to_stdout
-            "/dev/stdout"
-          else
-            "/dev/stdin"
+      bundle_file = "#{ENV["HOMEBREW_BUNDLE_FILE"]}"
```
Comment (colindean) [inline lib/bundle/brewfile.rb]:
```suggestion
      bundle_file = ENV["HOMEBREW_BUNDLE_FILE"]
```

I think I'd prefer this to be named `env_bundle_file` or the like to make its origin clear since it is not mutated later.

**Comment by colindean** [comment]
Thinking about how I'd use this _reasonably˚_ myself, I see this pattern:

* `HOMEBREW_BUNDLE_FILE` is set to `~/my_dotfiles/Brewfile.work` on my work machine.
* `HOMEBREW_BUNDLE_FILE` is set to `~/my_dotfiles/Brewfile.home` on my home machine.
* Project has a `Brewfile` in the root of the repo ➡️ Use automatic `./Brewfile`
* Project has a `Brewfile` but it's not in the root directory of the repo ➡️ specify `--file path/to/Brewfile`
* ❌ because then I can't use `--file` to specify when I'm using the envvar.

This feels like a reasonable pattern for me, esp. if we were to implement #365 adding a CLI for Brewfile management. As written, this usage pattern would break because of `--file` and envvar together erroring instead of the CLI argument taking precedence.

I think it makes sense for the envvar and `--global` to blow up when used together. _I_ don't see a usage pattern where someone would _want_ to use both, which would be maintaining _two_ active "master" Brewfiles per machine. 

_˚ I could get real creative but 90% of our users won't._

**Comment by MikeMcQuaid** [comment]
Great work here @aramprice and a great idea!

> * ❌ because then I can't use `--file` to specify when I'm using the envvar.

Agreed that it'd be nice to handle the `--file` case. I'm also on the fence about `--global`.

File: lib/bundle/brewfile.rb (hunk 1 above)
Comment (aramprice) [reply lib/bundle/brewfile.rb]: I like the rename to `env_bundle_file`.

I added the interpolation to prevent having to distinguish between `nil` vs `''` (empty string). That being said I didn't look into whether `ENV['FOO']` would return `nil` in cases where `FOO` was not set.

**Comment by aramprice** [comment]
I think relaxing what I have to allow `HOMEBREW_BUNDLE_FILE` to co-exist with `--file`, where `--file` wins is reasonable.

I suspect this will be a more invasive change. Would all you prefer that I PR some structural refactors to and then PR a feature change?

I'm also curious about the build failure. Looking at the logs it seems to be rubocop issues but when I run `brew style homebrew/bundle` there are no warnings (see below). Is there an update to the style checker that I need to do?

```
→ brew style homebrew/bundle

20 files inspected, no offenses detected

10:14:32 laptop in ~/workspace/homebrew-bundle
```

File: lib/bundle/brewfile.rb (hunk 1 above)
Comment (aramprice) [reply lib/bundle/brewfile.rb]: I could implement this as `ENV.fetch("HOMEBREW_BUNDLE_FILE", "")` which might be more clear. It would also make it easier to mock, as well as test for `nil` vs `""` (empty string).

**Comment by aramprice** [comment]
Reminder that currently if both `--global` and `--file` are specified the `--global` flag wins. This is the existing behavior.

File: lib/bundle/brewfile.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply lib/bundle/brewfile.rb]: > I added the interpolation to prevent having to distinguish between nil vs '' (empty string).

Comment (MikeMcQuaid) [reply lib/bundle/brewfile.rb]: > I added the interpolation to prevent having to distinguish between nil vs '' (empty string).

Can use `.blank?` instead.

> That being said I didn't look into whether `ENV['FOO']` would return `nil` in cases where `FOO` was not set.

It does.

Comment (aramprice) [reply lib/bundle/brewfile.rb]: Sad days `undefined method `blank?' for "":String`. I'll try `.fetch()`

**Comment by MikeMcQuaid** [comment]
> Would all you prefer that I PR some structural refactors to and then PR a feature change?

Fine either way with me, whatever you'd rather.

> I'm also curious about the build failure. Looking at the logs it seems to be rubocop issues but when I run `brew style homebrew/bundle` there are no warnings (see below). Is there an update to the style checker that I need to do?

That's not quite sufficient. Running `bundle exec rubocop` from the repository root should be.

> Reminder that currently if both `--global` and `--file` are specified the `--global` flag wins. This is the existing behavior.

Perhaps `--file` should or it should error? I'm open to either option if it makes the PR easier/better or the behaviour make more sense.

Thanks again @aramprice!

**Comment by aramprice** [comment]
I tried running `bundle exec rubocop` and got:

```
Inspecting 60 files
CCCCCCCCCCCWCCCCCCCWWCCCC..C..CCCCWCCWCWCWWWCCCWCCWWCCCWCCWC

Offenses:
# ... SNIPPED....
60 files inspected, 2315 offenses detected
```

The CI system is reporting:
```
2019-06-12T17:23:48.2726270Z More information: https://github.com/rubocop-hq/rubocop-rails
2019-06-12T17:23:55.0679970Z == /usr/local/Homebrew/Library/Taps/homebrew/homebrew-bundle/spec/bundle/brewfile_spec.rb ==
2019-06-12T17:23:55.0680760Z C: 81: 17: Space missing to the left of {.
2019-06-12T17:23:55.0687980Z C: 89: 17: Space missing to the left of {.
2019-06-12T17:23:55.0688340Z 
2019-06-12T17:23:55.0688650Z 21 files inspected, 2 offenses detected
```

but doesn't print out what command is actually running.

I can probably fix errors after they're reported by CI but it would be nice to be able to check locally.

**Comment by MikeMcQuaid** [comment]
https://github.com/Homebrew/homebrew-bundle/blob/master/azure-pipelines.yml details what CI is running.

**Comment by jacobbednarz** [comment]
The logic here sounds reasonable however it did take me a little while to get my head around the precedence stuff. I wonder if it's worth explicitly documenting the order and what overrides what to ensure we maintain it? Ideally some test cases can help be useful but documentation is probably a better first step.

**Comment by aramprice** [comment]
In case folks have troubles in the future with `brew style homebrew/bundle` the trick is to run `brew update-reset`.

**Comment by aramprice** [comment]
@jacobbednarz I think I've made the code more clear at the expense of readability of the PR.

The cascade is as follows:
- `--global` first
  - `--global` + `HOMEBREW_BUNDLE_FILE` results in an error
- `--file` trumps `HOMEBREW_BUNDLE_FILE`; no error
- `HOMEBREW_BUNDLE_FILE` if there are no other flags
- otherwise `${PWD}/Brewfile`

The function `handle_file_value` should probably be private but I wasn't sure how this interacted with `module_function`.

File: lib/bundle/brewfile.rb [hunk 2]
```diff
@@ -2,22 +2,22 @@
 
 module Bundle
   module Brewfile
+    ENV_KEY = "HOMEBREW_BUNDLE_FILE"
     module_function
 
     def path(dash_writes_to_stdout: false)
+      env_bundle_file = ENV.fetch(ENV_KEY, "")
```
Comment (colindean) [inline lib/bundle/brewfile.rb]: 👍

File: lib/bundle/brewfile.rb [hunk 3]
```diff
@@ -2,22 +2,22 @@
...
-        if filename == "-"
-          filename = if dash_writes_to_stdout
-            "/dev/stdout"
-          else
-            "/dev/stdin"
-          end
-        end
-        filename ||= "Brewfile"
+      elsif ARGV.include?("--file")
+        filename = handle_file_value(ARGV.value("file"), dash_writes_to_stdout)
```
Comment (colindean) [inline lib/bundle/brewfile.rb]: Looks like you could move the call to `ARGV.value("file")` here into the `handle_file_value` method since that method isn't being invoked elsewhere.

Comment (aramprice) [reply lib/bundle/brewfile.rb]: That makes sense. I chose to keep `ARGV` access within `#path` so that all access to `ARGV` could be seen at a glance in a single method.

File: lib/bundle/brewfile.rb [hunk 4]
```diff
@@ -2,22 +2,22 @@
 
 module Bundle
   module Brewfile
+    ENV_KEY = "HOMEBREW_BUNDLE_FILE"
```
Comment (MikeMcQuaid) [inline lib/bundle/brewfile.rb]: Don't think this is worth making a constant for a single use in this file.

File: lib/bundle/brewfile.rb (hunk 2 above)
Comment (MikeMcQuaid) [inline lib/bundle/brewfile.rb]: Use `ENV["HOMEBREW_BUNDLE_FILE"]`

File: lib/bundle/brewfile.rb [hunk 5]
```diff
@@ -2,22 +2,22 @@
...
     module_function
 
     def path(dash_writes_to_stdout: false)
+      env_bundle_file = ENV.fetch(ENV_KEY, "")
+
       if ARGV.include?("--global")
+        raise "'#{ENV_KEY}' can not be specified with '--global'" unless env_bundle_file.empty?
```
Comment (MikeMcQuaid) [inline lib/bundle/brewfile.rb]: Again: use `.blank?` please.

File: lib/bundle/brewfile.rb (hunk 3 above)
Comment (MikeMcQuaid) [reply lib/bundle/brewfile.rb]: I'm fine with either approach.

File: lib/bundle/brewfile.rb [hunk 6]
```diff
@@ -2,22 +2,22 @@
...
-            "/dev/stdout"
-          else
-            "/dev/stdin"
-          end
-        end
-        filename ||= "Brewfile"
+      elsif ARGV.include?("--file")
+        filename = handle_file_value(ARGV.value("file"), dash_writes_to_stdout)
         Pathname.new(filename).expand_path(Dir.pwd)
+      elsif !env_bundle_file.empty?
```
Comment (MikeMcQuaid) [inline lib/bundle/brewfile.rb]: `elsif env_bundle_file.present?`

File: lib/bundle/brewfile.rb [hunk 7]
```diff
@@ -2,22 +2,22 @@
 
 module Bundle
   module Brewfile
+    ENV_KEY = "HOMEBREW_BUNDLE_FILE"
     module_function
 
     def path(dash_writes_to_stdout: false)
+      env_bundle_file = ENV.fetch(ENV_KEY, "")
+
       if ARGV.include?("--global")
```
Comment (MikeMcQuaid) [inline lib/bundle/brewfile.rb]: Could assign all the results to a variable and then `Pathname.new` that variable to avoid `Pathname.new` being repeated in each case.

File: lib/bundle/brewfile.rb (hunk 5 above)
Comment (aramprice) [reply lib/bundle/brewfile.rb]: `.blank?` doesn't work.

```
ENV["HOME"].blank?
NoMethodError: undefined method `blank?' for "/Users/aram":String
	from (irb):2
	from /usr/bin/irb:11:in `<main>'
```

File: lib/bundle/brewfile.rb (hunk 2 above)
Comment (aramprice) [reply lib/bundle/brewfile.rb]: This postpones distinguishing between `nil` and `""` (empty string). I'm using `.fetch` to ensure that subsequent access to `env_bundle_file` is guaranteed to be operating on `String`.

File: lib/bundle/brewfile.rb (hunk 7 above)
Comment (aramprice) [reply lib/bundle/brewfile.rb]: That makes sense. It looks like the `.expand_path(Dir.pwd)` can safely be added to `Pathname.new("#{ENV["HOME"]}/.Brewfile")`.

File: lib/bundle/brewfile.rb (hunk 4 above)
Comment (aramprice) [reply lib/bundle/brewfile.rb]: For one I agree. There are two so maybe its arguable.

File: lib/bundle/brewfile.rb (hunk 5 above)
Comment (MikeMcQuaid) [reply lib/bundle/brewfile.rb]: Use `brew irb` instead. It's a ActiveSupport method included by Homebrew.

File: lib/bundle/brewfile.rb (hunk 6 above)
Comment (aramprice) [reply lib/bundle/brewfile.rb]:
```
→ bundle exec irb
irb(main):001:0> "".present?
NoMethodError: undefined method `present?' for "":String
# ...
```

Comment (jacobbednarz) [reply lib/bundle/brewfile.rb]: @aramprice You'll need `brew irb` as @MikeMcQuaid mentioned in https://github.com/Homebrew/homebrew-bundle/pull/502#discussion_r293581893 as it's included by Homebrew, not just bundler like you're attempting here.

Comment (aramprice) [reply lib/bundle/brewfile.rb]: These methods are present when I run `brew irb` but tests fail when I run `bundle exec rake`. This appears to be what is [run in CI](https://github.com/Homebrew/homebrew-bundle/blob/master/azure-pipelines.yml#L13) so my expectation is that CI will not pass with `.blank?` or `.present?`

File: lib/bundle/brewfile.rb (hunk 5 above)
Comment (aramprice) [reply lib/bundle/brewfile.rb]: `.blank?` works when I run `brew irb` but tests fail when I run `bundle exec rake`.

File: lib/bundle/brewfile.rb (hunk 1 above)
Comment (aramprice) [reply lib/bundle/brewfile.rb]: See https://github.com/Homebrew/homebrew-bundle/pull/502#discussion_r293596295 regarding problems using `.blank?` or `.present?`. If these can be used I'm happy to change from `ENV.fetch()` to `ENV[]`.

File: lib/bundle/brewfile.rb (hunk 6 above)
Comment (MikeMcQuaid) [reply lib/bundle/brewfile.rb]: Yes, it will need mocked in the unit tests.

File: lib/bundle/brewfile.rb [hunk 8]
```diff
@@ -2,29 +2,39 @@
 
 module Bundle
   module Brewfile
+    BUNDLE_FILE_ENV_KEY = "HOMEBREW_BUNDLE_FILE"
```
Comment (MikeMcQuaid) [inline lib/bundle/brewfile.rb]: Again, please hardcode this: https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)

File: lib/bundle/brewfile.rb [hunk 9]
```diff
@@ -2,29 +2,39 @@
...
-      else
-        filename = ARGV.value("file")
-        if filename == "-"
-          filename = if dash_writes_to_stdout
-            "/dev/stdout"
+      env_bundle_file = ENV[BUNDLE_FILE_ENV_KEY]
+
+      filename =
+          if ARGV.include?("--global")
+            raise "'#{BUNDLE_FILE_ENV_KEY}' can not be specified with '--global'" unless env_bundle_file.blank?
```
Comment (MikeMcQuaid) [inline lib/bundle/brewfile.rb]: `if env_bundle_file.present?` to avoid double negative.

File: lib/bundle/brewfile.rb [hunk 10]
```diff
@@ -2,29 +2,39 @@
...
     end
 
     def read
       Brewfile.path.read
     rescue Errno::ENOENT
       raise "No Brewfile found"
     end
+
+    def handle_file_value(filename, dash_writes_to_stdout)
+      if filename == "-"
```
Comment (MikeMcQuaid) [inline lib/bundle/brewfile.rb]: `if filename != "-"` and then could do `elsif dash_writes_to_stdout` and `else` which would make the function a little more readable.

File: spec/bundle/brewfile_spec.rb [hunk 11]
```diff
@@ -4,49 +4,190 @@
 
 describe Bundle::Brewfile do
   describe "path" do
-    context "when `--file` is passed" do
-      before do
-        allow(ARGV).to receive(:include?).with("--global").and_return(false)
-        allow(ARGV).to receive(:value).with("file").and_return(file_value)
+    # monkeypatch Object so it behaves as if ActiveSupport were present
+    class Object
```
Comment (MikeMcQuaid) [inline spec/bundle/brewfile_spec.rb]: Either put this in `spec/stub` or use `allow` or `expect`

File: lib/bundle/brewfile.rb (hunk 8 above)
Comment (aramprice) [reply lib/bundle/brewfile.rb]: as you like

**Comment by MikeMcQuaid** [comment]
Thanks for all this @aramprice and sorry for all the back and forth. Great work here, you rock!

File: lib/bundle/brewfile.rb [hunk 12]
```diff
@@ -5,26 +5,37 @@ module Brewfile
...
-        if filename == "-"
-          filename = if dash_writes_to_stdout
-            "/dev/stdout"
+      env_bundle_file = ENV["HOMEBREW_BUNDLE_FILE"]
+
+      filename =
+          if ARGV.include?("--global")
+            raise "'HOMEBREW_BUNDLE_FILE' can not be specified with '--global'" if env_bundle_file.present?
+            "#{ENV["HOME"]}/.Brewfile"
+          elsif ARGV.include?("--file")
```
Comment (andyli) [inline lib/bundle/brewfile.rb]: When using `brew bundle --file=path/to/Brewfile`, `ARGV` will be `["--file=path/to/Brewfile"]`. So, the check of `ARGV.include?("--file")` is wrong.

**Comment by colindean** [comment]
Thank you for your hard work on this, @aramprice! I appreciate you!

File: lib/bundle/brewfile.rb (hunk 12 above)
Comment (aramprice) [reply lib/bundle/brewfile.rb]: Thanks for pointing this out. It seems like this might need to be adjusted in the mainline now that the PR has been closed.

**Comment by danielbayley** [comment]
> Great work here @aramprice and a great idea!

@MikeMcQuaid What changed between [then](https://github.com/Homebrew/homebrew-bundle/pull/261) (#261) and now? 😅

Anyway, glad this was finally allowed… nice one @aramprice!

**Comment by MikeMcQuaid** [comment]
@danielbayley 2.5 years and a bunch more people asking for it. I realise it must be frustrating but it's general good product practice to want to hear demand from more users before implementing new features.

**Comment by danielbayley** [comment]
> I realise it must be frustrating

@MikeMcQuaid Not a big deal 😉
